Do not trigger url rewrites re-generation for all store views

ihor-sviziev opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
During creating a new website, Store Group and Store view using `php bin/magento setup:upgrade` - we noticed that some URL rewrites for categories on **different** stores were re-generated in DB, that's absolutely not obvious and for sure not expected behavior. Such behavior, especially for us changed URLs for some categories _w/o creating any redirect from old -> new URL_ (for some reason, we had incorrect url_path values in DB, that's another story). 
As a result of this issue - some advertised categories started returning 404 pages, and we basically started burning our money.

#### Root cause analyze
Investigation showed, that `app/code/Magento/CatalogUrlRewrite/Model/Category/Plugin/Store/Group.php` plugin was executed due to 2 step saving of group:
- at first, we just creating a new group
- later we're assigning it to a website
https://github.com/magento/magento2/blob/6729b6e01368248abc33300208eb292c95050203/app/code/Magento/Store/Model/Config/Importer/Processor/Create.php#L165-L195

As a store group still don't have any store IDs assigned (the store will be created on the next step) - `$group->getStoreIds()` returns an empty array, we're setting this value to the category and trying to generate URL rewrites
https://github.com/magento/magento2/blob/a2b93809864fa6dd645d44a5c818266459f53b95/app/code/Magento/CatalogUrlRewrite/Model/Category/Plugin/Store/Group.php#L155-L157

Then it goes to this method, which tries to fetch a list of store Ids:
https://github.com/magento/magento2/blob/c051134a4b2c837f798cc8f66b9ad4c208310363/app/code/Magento/CatalogUrlRewrite/Model/CategoryUrlRewriteGenerator.php#L116-L143

Here we don't have any saved `store_ids`, so it fetches the list of stores where this category used as a root.
https://github.com/magento/magento2/blob/68c7282c414916d5929528f472a3f064422aa12d/app/code/Magento/Catalog/Model/Category.php#L534-L570

As a result - we're re-generating category URL rewrites for **all store views with the same root category ID**. 
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
1. Install magento, so you'll have 1 website, 1 store group, 1 store view
2. Create some categories structure
3. Run `php bin/magento app:config:dump scopes` to dump all the stores data into the config.php
4. Create a DB dump
5. Go to Admin, and create a new website, store group and a store view (re-use the same root category for that)
6. Run `php bin/magento app:config:dump scopes` to dump all the stores data into the config.php, save the file
7. Apply the DB dump from the step 4 in order to emulate the case, when we didn't have 2nd website, store group, etc
8. Run `php bin/magento setup:upgrade` or `php bin/magento app:config:import` to create a new website, store group, store view 
9. Compare data in the `url_rewrites` table with a backup file for the store 1.

**Expected result:**
✔ URL rewrites for store 1 shouldn't be touched when we creating a new website / store group / store view

**Actual result:**
❌ URL rewrites for store 1 are re-generated when we creating a new website / store group / store view
### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#32954: Do not trigger url rewrites re-generation for all store views

ihor-sviziev committed:
Do not trigger url rewrites re-generation for all store views

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run Functional Tests EE, WebAPI Tests

gabrieldagama commented:
Hi @ihor-sviziev, thanks for contributing! Can you provide more details on the impact of this fix? Thanks!

ihor-sviziev committed:
Do not trigger url rewrites re-generation for all store views

Fix unit tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
Hi @gabrieldagama,
I updated the PR description

ihor-sviziev commented:
@magento run Functional Tests B2B

ihor-sviziev commented:
@magento run Functional Tests B2B

gabrieldagama commented:
@magento run Functional Tests B2B

Den4ik commented:
@magento create issue

Den4ik commented:
@ihor-sviziev Thanks for collaboration and detailed description.
Do we really need save group 2 times at https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Store/Model/Config/Importer/Processor/Create.php#L187-L193 ?

ihor-sviziev commented:
@Den4ik, in theory, we might change it, but I don't know why it was done in this way and wonder if it would cause some regression. Also, potentially, there might be similar code in some other piece or custom extension with the same behavior. My current changes fix all cases at once, so should cover all cases

Den4ik commented:
@ihor-sviziev May be we could little bit research and if it's possible reduce count of save calls that will have a positive effect on performance?

ihor-sviziev commented:
@Den4ik, in general, it's a good idea to do such research and get rid of such pieces. But in the case of fixing this issue - that's not relevant  - I won't break any functionality.

Den4ik reviewed (approved):
:heavy_check_mark: Approve. All looks good for me.
As we discussed in Slack research should be done as separate work.

magento-engcom-team commented:
Hi @Den4ik, thank you for the review.
[ENGCOM-9074](https://jira.corp.magento.com/browse/ENGCOM-9074) has been created to process this Pull Request
:eight_spoked_asterisk: @Den4ik, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

engcom-Alfa commented:
Hi @ihor-sviziev , @Den4ik 
Thanks for the collaboration & contribution!

:x: QA not Passed

_URL Re-Write regenerates for all the store views_

**Execution steps followed as below**

1. Installed a Magento freshly having PR changes, so by default l have 1 website, 1 store group, 1 store view.
2. Created a 3 categories inside default category and one product for better analysis. Noted the below URL Re Writes table and the data of the table is below:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/MANJUN~1.KS/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/MANJUN~1.KS/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:10.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Arial;
	mso-generic-font-family:auto;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:white;
	mso-pattern:white none;}
.xl64
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:whitesmoke;
	mso-pattern:whitesmoke none;}
.xl65
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:#E5F7FE;
	mso-pattern:#E5F7FE none;}
.xl66
	{border:.5pt solid windowtext;}
-->
</style>
</head>

<body link="#1155CC" vlink="#1155CC">



ID | Store View | Request Path | Target Path | Redirect Type
-- | -- | -- | -- | --
1 | Main Website      Main Website Store      Default Store View | no-route | cms/page/view/page_id/1 | No
2 | Main Website      Main Website Store      Default Store View | home | cms/page/view/page_id/2 | No
3 | Main Website      Main Website Store      Default Store View | enable-cookies | cms/page/view/page_id/3 | No
4 | Main Website      Main Website Store      Default Store View | privacy-policy-cookie-restriction-mode | cms/page/view/page_id/4 | No
5 | Main Website      Main Website Store      Default Store View | mks1.html | catalog/category/view/id/3 | No
6 | Main Website      Main Website Store      Default Store View | mks1/mks2.html | catalog/category/view/id/4 | No
7 | Main Website      Main Website Store      Default Store View | mks1/mks3.html | catalog/category/view/id/5 | No
8 | Main Website      Main Website Store      Default Store View | mksprod1.html | catalog/product/view/id/1 | No
9 | Main Website      Main Website Store      Default Store View | mks1/mks3/mksprod1.html | catalog/product/view/id/1/category/5 | No
10 | Main Website      Main Website Store      Default Store View | mks1/mksprod1.html | catalog/product/view/id/1/category/3 | No



</body>

</html>


3. Ran the command _php bin/magento app:config:dump scopes_
4. Created a DB dump
5. Created a new store with new website and store view, setting the same default category. Checked the new store domain is able to open in the front end as shown in the below screenshot.
![image](https://user-images.githubusercontent.com/51679138/134649520-fe43259a-3e12-4b8b-8736-4124457778e2.png)
Where the default main store had the domain with the url _local.prurlrewriteafter.com_, and now the additional store is with _local.prurlrewriteafter.fr_. Both stores front end are working fine.

6. Noted the updated URL Re Write table again and the relevant table data is below
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/MANJUN~1.KS/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/MANJUN~1.KS/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:10.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Arial;
	mso-generic-font-family:auto;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{border:.5pt solid windowtext;}
.xl64
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:white;
	mso-pattern:white none;}
.xl65
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:whitesmoke;
	mso-pattern:whitesmoke none;}
.xl66
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:#E5F7FE;
	mso-pattern:#E5F7FE none;}
-->
</style>
</head>

<body link="#1155CC" vlink="#1155CC">



ID | Store View | Request Path | Target Path | Redirect Type
-- | -- | -- | -- | --
1 | Main Website      Main Website Store      Default Store View | no-route | cms/page/view/page_id/1 | No
2 | Main Website      Main Website Store      Default Store View | home | cms/page/view/page_id/2 | No
3 | Main Website      Main Website Store      Default Store View | enable-cookies | cms/page/view/page_id/3 | No
4 | Main Website      Main Website Store      Default Store View | privacy-policy-cookie-restriction-mode | cms/page/view/page_id/4 | No
5 | Main Website      Main Website Store      Default Store View | mks1.html | catalog/category/view/id/3 | No
6 | Main Website      Main Website Store      Default Store View | mks1/mks2.html | catalog/category/view/id/4 | No
7 | Main Website      Main Website Store      Default Store View | mks1/mks3.html | catalog/category/view/id/5 | No
8 | Main Website      Main Website Store      Default Store View | mksprod1.html | catalog/product/view/id/1 | No
9 | Main Website      Main Website Store      Default Store View | mks1/mks3/mksprod1.html | catalog/product/view/id/1/category/5 | No
10 | Main Website      Main Website Store      Default Store View | mks1/mksprod1.html | catalog/product/view/id/1/category/3 | No
11 | FrenchWebsite      FrenchStore      FrenchStoreView | mks1.html | catalog/category/view/id/3 | No
12 | FrenchWebsite      FrenchStore      FrenchStoreView | mks1/mks2.html | catalog/category/view/id/4 | No
13 | FrenchWebsite      FrenchStore      FrenchStoreView | mks1/mks3.html | catalog/category/view/id/5 | No
14 | FrenchWebsite      FrenchStore      FrenchStoreView | no-route | cms/page/view/page_id/1 | No
15 | FrenchWebsite      FrenchStore      FrenchStoreView | home | cms/page/view/page_id/2 | No
16 | FrenchWebsite      FrenchStore      FrenchStoreView | enable-cookies | cms/page/view/page_id/3 | No
17 | FrenchWebsite      FrenchStore      FrenchStoreView | privacy-policy-cookie-restriction-mode | cms/page/view/page_id/4 | No



</body>

</html>

7. Ran _php bin/magento app:config:dump scopes_
8. Applied the DB dump back to Magento instance from the step 4.
9. Ran php bin/magento setup:upgrade and followed by static deploy and re indexed as well.
10. Noted the updated URL Re Write table again and the relevant table data is below
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/MANJUN~1.KS/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/MANJUN~1.KS/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:10.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Arial;
	mso-generic-font-family:auto;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{border:.5pt solid windowtext;}
.xl64
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:white;
	mso-pattern:white none;}
.xl65
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:whitesmoke;
	mso-pattern:whitesmoke none;}
.xl66
	{color:#303030;
	font-family:"\0022Open Sans\0022";
	mso-generic-font-family:auto;
	mso-font-charset:0;
	border:.5pt solid windowtext;
	background:#E5F7FE;
	mso-pattern:#E5F7FE none;}
-->
</style>
</head>

<body link="#1155CC" vlink="#1155CC">



ID | Store View | Request Path | Target Path | Redirect Type
-- | -- | -- | -- | --
1 | Main Website      Main Website Store      Default Store View | no-route | cms/page/view/page_id/1 | No
2 | Main Website      Main Website Store      Default Store View | home | cms/page/view/page_id/2 | No
3 | Main Website      Main Website Store      Default Store View | enable-cookies | cms/page/view/page_id/3 | No
4 | Main Website      Main Website Store      Default Store View | privacy-policy-cookie-restriction-mode | cms/page/view/page_id/4 | No
8 | Main Website      Main Website Store      Default Store View | mksprod1.html | catalog/product/view/id/1 | No
9 | Main Website      Main Website Store      Default Store View | mks1/mks3/mksprod1.html | catalog/product/view/id/1/category/5 | No
10 | Main Website      Main Website Store      Default Store View | mks1/mksprod1.html | catalog/product/view/id/1/category/3 | No
15 | Main Website      Main Website Store      Default Store View | mks1.html | catalog/category/view/id/3 | No
16 | Main Website      Main Website Store      Default Store View | mks1/mks2.html | catalog/category/view/id/4 | No
17 | Main Website      Main Website Store      Default Store View | mks1/mks3.html | catalog/category/view/id/5 | No
21 | FrenchWebsite      FrenchStore      FrenchStoreView | mks1.html | catalog/category/view/id/3 | No
22 | FrenchWebsite      FrenchStore      FrenchStoreView | mks1/mks2.html | catalog/category/view/id/4 | No
23 | FrenchWebsite      FrenchStore      FrenchStoreView | mks1/mks3.html | catalog/category/view/id/5 | No
24 | FrenchWebsite      FrenchStore      FrenchStoreView | no-route | cms/page/view/page_id/1 | No
25 | FrenchWebsite      FrenchStore      FrenchStoreView | home | cms/page/view/page_id/2 | No
26 | FrenchWebsite      FrenchStore      FrenchStoreView | enable-cookies | cms/page/view/page_id/3 | No
27 | FrenchWebsite      FrenchStore      FrenchStoreView | privacy-policy-cookie-restriction-mode | cms/page/view/page_id/4 | No



</body>

</html>



**:heavy_check_mark: Expected result After Fix** 
- Along with the default store's front end, the newly created front end should also work fine!
- The URL Rewrite data should not be re generated

**:x: Actual result After Fix** 
- If we hit the store's front end domain, redirecting to main default store domain.
- URL Rewrite data updated with newly generated record IDs.

engcom-Alfa reviewed (changes requested):
Based on the above comment, moving it back to to _Request Changes_

ihor-sviziev commented:
Please ping me in slack for further details

On Mon, 11 Oct 2021 at 17:55 Indrani ***@***.***> wrote:

> Hi @ihor-sviziev <https://github.com/ihor-sviziev>,
>
> Thank you for your contribution!
>
> I am trying to understand and reproduce this issue on my local. Followed
> below steps but I am getting the same results on 2.4 develop and PR branch.
>
>    1. Setup fresh magento instance
>    2. Added 2 new sub categories under Default Category. Created 2
>    products under one sub category and 1 in other. Eg.(Electronics and Cloths
>    are sub categories and lamp, geyser are products under Electronics and
>    t-shirt product is under cloths)
>    3. Checked url_rewrite table: total 12 records were present.
>    4. Execute php bin/magento app:config:dump scopes command
>    5. Took Db dump at this stage
>    6. Added one website, store and view against default category
>    7. Checked the url_rewrite table again. Now total 18 records were
>    present
>    8. Executed php bin/magento app:config:dump scopes again
>    9. Applied the DB dump back to Magento instance from the step 5.
>    10. Executed php bin/magento setup:upgrade, At this, it asked about
>    “do you want to continue with creating the website” of step 7 , I have
>    opted for Yes option
>    11. Tried to browse the site again, its accessible and checked
>    url_rewrite table again. Its having 18 records
>    12. Executed php bin/magento app:config:import cmd : Got nothing to
>    import message here
>
> On both the branches ie. on 2.4 develop and on this PR branch,
> url_rewrite_ids of few records are getting changed after step 9 on, so can
> you please confirm once and give us some more information on this.
>
> Thank you in advance!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magento/magento2/pull/32808#issuecomment-940104457>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOJOUMHRSG6W4O6ADQPDPLUGL3H5ANCNFSM43KV2QRQ>
> .
>
